### PR TITLE
MAPSAND-683 Make Source constructors private

### PR DESCRIPTION
**Pull request by paulrenn67**
### Summary of changes

Source constructor visibility changed to private, users should use Builder class instead.

### User impact (optional)

Usage of e.g. GeoJsonSource(builder) should be changed to GeoJsonSource.Builder().build()

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [x] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

**Comment by ank27** [comment]
make sure to add tests and changelog :)

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 1]
```diff
@@ -34,23 +33,14 @@ import com.mapbox.maps.logW
  * @see [The online documentation](https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/#geojson)
  *
  */
-class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
+class GeoJsonSource private constructor(builder: Builder) : Source(builder.sourceId) {
```
Comment (yunikkk) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: I'm having second thoughts about other Sources than GeoJsonSource. While for this one it's definitely to call `GeoJsonSource(builder)` on the user end other Sources just return `Source(builder)` in their `Builder.build()` definition, e.g. https://github.com/mapbox/mapbox-maps-android/blob/f6bd9832e16405742af74faacfb7f817a9ada03c/extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/ImageSource.kt#L140. 
I can imagine there might be a small change someone is using those constructors. To reduce risk of unneeded runtime crashes could we maybe only make private constructor for the GeoJsonSource? And just deprecate rest of Sources?

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: I think all the Souce classes should have a consistent API

Comment (yunikkk) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: I see your point about consistency. At the same time it still concerns me that we introduce potential crash for the users who might use sources with perfectly working constructors of Image/Raster/RasterDem/Vector sources. While for GeoJsonSource this constructor is buggy, other sources are fine.

Comment (yunikkk) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: Actually I've just got another idea - use Deprecated annotation with level ERROR - this way it'll remain binary but not  source compatible, thus forcing to use another one @paulrenn67

Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: I think it make sense to deprecate and use another one, so we don't break user's code. 
Or we can treat it as a breaking change?

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:36]: Hmmm ok, we can go with the Deprecation approach for now, and remove in v11

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: @yunikkk @ank27 could you please review again, constructors now made `Deprecated`. Had to use `DeprecationLevel.WARNING` though, otherwise our own `Builder.build `gets compile error.

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 Mapbox welcomes participation and contributions from everyone.
 
 # main
+* Source constructor visibility changed to private, users should use Builder class instead. ([1975](https://github.com/mapbox/mapbox-maps-android/pull/1975))
```
Comment (ank27) [inline CHANGELOG.md:6]:
```suggestion
* Deprecate `Source` constructor invocation. Use `Builder` class instead. ([1975](https://github.com/mapbox/mapbox-maps-android/pull/1975))
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 3]
```diff
@@ -34,21 +33,19 @@ import com.mapbox.maps.logW
  * @see [The online documentation](https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/#geojson)
  *
  */
-class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
-  private val workerHandler by lazy {
-    Handler(workerThread.looper)
+class GeoJsonSource : Source {
+  @Deprecated("Use builder instead", level = DeprecationLevel.WARNING)
```
Comment (ank27) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:37]:
```suggestion
  @Deprecated("This method is deprecated. Use Builder instead", level = DeprecationLevel.WARNING)
```

**Comment by ank27** [comment]
Looks good in general. can you check the generated code, looks like there are some indentation issues

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 4]
```diff
@@ -649,11 +638,12 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
      *
      * @return the GeoJsonSource
      */
+    @SuppressWarnings("deprecation")
```
Comment (ank27) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:641]: No need to `SuppressWarning` imo.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 5]
```diff
@@ -34,21 +33,19 @@ import com.mapbox.maps.logW
...
  */
-class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
-  private val workerHandler by lazy {
-    Handler(workerThread.looper)
+class GeoJsonSource : Source {
+  @Deprecated("Use builder instead", level = DeprecationLevel.WARNING)
+  constructor(builder: Builder) : super(builder.sourceId) {
+      sourceProperties.putAll(builder.properties)
+      volatileSourceProperties.putAll(builder.volatileProperties)
+    initGeoJson = builder.geoJson
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:41]: smth with tabs/spaces

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 3 above)
Comment (yunikkk) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:37]: I'd suggest to use level Error to force users to use Builder. It'll still be source-incompatible change, breaking compilation but not runtime behaviour.

Comment (yunikkk) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:37]: E.g.

```suggestion
  @Deprecated("Use builder instead", level = DeprecationLevel.ERRROR)
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 1 above)
Comment (pengdev) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: I think we might overcomplicated things here, I'd consider it as a bug fix, using constructor with Builder inside it is clearly a misuse to me, and user should fix their code if they were using it that way. wdyt @yunikkk @ank27 @paulrenn67 ?

Comment (yunikkk) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: @pengdev it's still better to break only source compatibility in my opinion. Otherwise it's binary breaking and if maps minor version is updated - they'll get a crash. 
At the same time clearly wrong behaviour would be to create GeoJsonSource using constructor with builder, other sources may be perfectly fine built using it.

Comment (yunikkk) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: > Had to use DeprecationLevel.WARNING though, otherwise our own Builder.build gets compile error.

@paulrenn67 could you check with  `@Suppress("DEPRECATION_ERROR")` ? Seems it can ignore even the error

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: Ah maybe DEPRECATION_ERROR will do the trick.

Comment (pengdev) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: @paulrenn67 I don't think it's worth the hassle, as we are going to remove the private constructor in v11 (and we don't recommend to use constructor in v10 from the beginning).
Especially these classes are generated and is hard to maintain separately for GeoJsonSource from other sources.
I'm fine with deprecation if you think it's necessary, please avoid complex logic as we are going to v11 soon.

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: @pengdev the issue is that @yunikkk raised the question of whether we should change anything at all.  So either we implement the ticket to make constructors private, or we can just discard this PR and mark the ticket as "won't do".  I think "won't do" makes more sense now if we are going to make bigger changes in V11.

Comment (yunikkk) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: > @yunikkk raised the question of whether we should change anything at all.

My preferred solution is to use Deprecated with the Error level, this way we 
- forbid compiling if wrong constructor is used, forcing users to change 
- avoid crashes if maps version was bumped implicitly
The only problem could be that compiler won't allow ourselves to use constructor but if that's hacked with `@Suppress("DEPRECATION_ERROR")` I think there's no more flaws here?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/ImageSource.kt [hunk 6]
```diff
@@ -17,12 +17,6 @@ import java.util.*
  *
  */
 class ImageSource(builder: Builder) : Source(builder.sourceId) {
-
-  init {
-    sourceProperties.putAll(builder.properties)
-    volatileSourceProperties.putAll(builder.volatileProperties)
-  }
```
Comment (yunikkk) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/ImageSource.kt:24]: I guess that wasn't intentional?

Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/ImageSource.kt:24]: Hmmm true, that should be retained for Sources that are _not_ GeoJson. Thanks for noticing.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 1 above)
Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: Constructor changed to 
```
class GeoJsonSource : Source {
  @Deprecated("Use builder instead", level = DeprecationLevel.ERROR)
  constructor(builder: Builder) : super(builder.sourceId) {
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 4 above)
Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt]: Without `@Suppress("DEPRECATION_ERROR") `we get `compiler error Using 'constructor GeoJsonSource(GeoJsonSource.Builder)' is an error. Use builder instead`

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/RasterDemSource.kt [hunk 7]
```diff
@@ -157,7 +156,7 @@ class RasterDemSource(builder: Builder) : Source(builder.sourceId) {
      */
     get() {
       getPropertyValue<String?>("encoding")?.let {
-        return Encoding.valueOf(it.toUpperCase(Locale.US).replace('-', '_'))
```
Comment (pengdev) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/RasterDemSource.kt:160]: is this change intended?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 8]
```diff
@@ -86,11 +84,6 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
     }
   }
 
-  init {
-    sourceProperties.putAll(builder.properties)
-    volatileSourceProperties.putAll(builder.volatileProperties)
-  }
-
```
Comment (pengdev) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:93]: any reason we removed this init block?

Comment (pengdev) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:93]: ok, found it was moved to the constructor,

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/RasterDemSource.kt (hunk 7 above)
Comment (paulrenn67) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/RasterDemSource.kt:160]: `toUpperCase` is deprecated in favour of `uppercase`

**Review by pengdev — APPROVED** [review]
LGTM % remaining open comments.

**Comment by pengdev** [comment]
Closing as we did it in the internal repo.
